doc/ip-sentinel.man
.\" $Id: ip-sentinel.man,v 1.9 2003/09/09 16:15:59 ensc Exp $
.\" Copyright (C) 2002 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
.\"  
.\" Permission is granted to make and distribute verbatim copies of
.\" this manual provided the copyright notice and this permission notice
.\" are preserved on all copies.
.\"  
.\" Permission is granted to process this file through groff/nroff and related
.\" tools and print the results, provided the printed document carries copying
.\" permission notice identical to this one except for the removal of this
.\" paragraph (this paragraph not being relevant to the printed manual).
.\"  
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a 
.\" permission notice identical to this one.
.\"  
.\" Permission is granted to copy and distribute translations of this manual
.\" into another language, under the above conditions for modified versions,
.\" except that this permission notice may be stated in a translation
.\" approved by the Free Software Foundation
.\"  
.TH ip-sentinel 8 "Aug 28 2003" "@PACKAGE@ 0.7"
.\"
.\" ====================
.\"
.SH NAME
.PP
ip-sentinel \- keeps your IP space clean
.\"
.\" ====================
.\"
.SH SYNOPSIS
.B ip-sentinel
.RB [ \-\-user | \-u
.IR username ]
.RB [ \-\-group | \-g
.IR groupname ]
.RB [ \-\-chroot | \-r
.IR dir ]
.RB [ \-\-ipfile | \-i
.IR file ]
.RB [ \-\-pidfile | \-p
.IR file ]
.RB [ \-\-logfile | \-l
.IR file ]
.RB [ \-\-errfile | \-e
.IR file ]
.RB [ \-\-mac
.IR mac ]
.RB [ \-\-llmac
.IR mac ]
.RB [ \-\-direction
.IR dir ]
.RB [ \-\-nofork | \-n ]
.RB [ \-\-help | \-h ]
.RB [ \-\-version ]
.IR interface
.\"
.\" ====================
.\"
.SH DESCRIPTION
.B ip-sentinel
starts a daemon which tries to prevent unauthorized usage of IPs
within an ethernet broadcastdomain by answering ARP requests. After
receiving such a faked reply, the requesting party stores the MAC
address in its ARP-table and will send future packets to this invalid
MAC, rendering the IP unreachable.

After startup it opens its logfiles, goes into a chroot jail, drops
privileges and forks itself into background by default. The parameters
of these operations can by changed by the options described below.

.B Warning:
.\"
with an uncareful configuration, ip-sentinel can bring down the entire
network and your physical presence at the ip-sentinel machine and/or
the switch will be required to fix it. Some switches can show an
unexpected behavior are even crash when the special \fI802.*\fR
mac-addresses will be used.
.\"
.\" ====================
.\"
.SH OPTIONS
.TP
.BR \-\-user | \-u\ \fIusername\ \fRand\fB\ \-\-group | \-g\ \fIgroupname
After doing basical initializations, the program will drop its
privilegies by changing its user and group ID. The \fIusername\fR and
\fIgroupname\fR parameters determine these new IDs and must be an
alphanumeric username/groupname or a numeric uid/gid.

.B Attention:
.\"
When having a dietlibc-compiled version in a system which uses remote
NSS (e.g. LDAP, NIS) to make passwd-lookups, you will have to use
numeric IDs.
.\"
.TP
.BR \-\-chroot | \-r\ \fIdirname
Specifies the directory where the program shall go into and call
.B chroot()
after finishing its initialization. By default it is the homedirectory
of the user specified with the \fB\-\-user\fR option.
.TP
.BR \-\-ipfile | \-i\ \fIfilename
Sets the name of the configuration file which contains the list of
forbidden IPs. This name is relative to the chroot and is
.BR ips.cfg
by default. This file will be reread automatically after changes.
.TP
.BR \-\-pidfile | \-p\ \fIfilename
Sets the file where the pid of the daemon will be written into. This
filename is relative to the root-filesystem and is
.BR /var/run/ip\-sentinel.run
by default.
.TP
.BR \-\-logfile | \-l\ \fIfilename
Sets the file where normal operations will be logged. Normal
operations are e.g. an ARP-reply for a forbidden IP. This filename is
relative to the root-filesystem and
.BR /var/log/ip-sentinel.out
by default.
.TP
.BR \-\-errfile | \-e\ \fIfilename
Sets the file where error-conditions will be logged. Error-conditions
are e.g. detected DOS attacks (bombing the ip-sentinel with
ARP-requests) or errors in the configuration file. This filename is
relative to the root-filesystem and
.BR /var/log/ip-sentinel.err
by default.
.TP
.BR \-\-nofork | \-n
Prevents daemon from going into background after initialization. This
option may be useful when using advanced init-concepts like minit
(http://www.fefe.de/minit/).
.TP
.BR \-\-direction\ \fIdir
ip-sentinel can answer both requests asking for an intruder's address,
and requests coming from an intruder asking for official
addresses. This parameter specifies which requests shall be
answered. Possible values are \fIFROM\fR (requests from intruders),
\fITO\fR (requests to intruders) and \fIBOTH\fR. The current default
value is \fITO\fR, but will probably change to \fIBOTH\fR in future
versions.
.TP
.BR \-\-mac\ \fImac
Use \fImac\fR as the default faked mac-address; see below for possible
values. This parameter can be overriden in the ipfile for each
questionable IP. The default value is \fIRANDOM\fR.
.TP
.BR \-\-llmac\ \fImac
Specifies the mac-address which will be used in the link-level headers
when answering requests \fBfrom\fR intruders. Additionally to the
values described below, ip-sentinel understands the \fISAME\fR string,
which means that the address from the ARP-header will be used. When
not using the \fIFROM\fR or \fIBOTH\fR directions, the assigned value
will not have any effect. The default is \fILOCAL\fR.
.TP
.BR \-\-help | \-h
Prints help message and exits.
.TP
.BR \-\-version
Prints the version of the program and exits.
.TP
.I interface
Specifies the network-interface where ip-sentinel listens for
ARP-requests and sends ARP-replies. It is a name like "eth0"; aliases
like "eth0:1" are not possible.
.\"
.\" ====================
.\"
.SH CONFIGURATION FILE
The location of the configfile is relatively to the choosen
chroot-directory. Its default is determined at compilation time but it
can be overriden by the \fB\-\-ipfile\fR option.

The content of this file has the following syntax:
.\"
.TP
.IR  ip\  [ mac ]
Specifies an IP which shall be blocked. "Blocked" means that
ARP-replies will be generated telling that the IP address \fIip\fR has
the MAC \fImac\fR. When the MAC parameter is not given, a random one
will be choosen.

.B Attention:
.\"
Assigning a MAC in a switched environment will not have the desired
effect, because the switch will be confused when receiving ARP-replies
for the same IP on different ports (this one of the bad-host and this
one of the ip-sentinel host).

Examples are: "192.168.42.23  11:22:33:44:55:66" or "10.0.0.0".
.TP
.IB net / mask\ \fR[ \fImac\fR]
Specifies that IPs of an entiry network shall be blocked. See above
for the meaning of "blocked". It is possible to specify a MAC address;
else a random one will be generated.

Examples are: "169.0.0.0/8", "192.168.8.15/255.255.65.31" or "192.168.23.42/26 a:b:c:d:e:f".
.TP
.BI ! ip
The given IP address \fIip\fR will be ignored. By default, any not
specified address will be ignored but when having blocked netmasks it
may be usefully to allow certain IPs.

An example is: "!192.168.23.42"
.TP
.BI ! net / mask
Tells that IPs of the given network shall be ignored.

Examples are: "!192.168.1.0/255.255.255.0" or "!0.0.0.0/0" (the
default).
.TP
.BI #\  comment\ \fR(and\ \fRempty\ \fRlines)
A comment; will be ignored

.PP
.\"
To be switch-friendly, there are only a few random MACs possible which
are having the format "de:ad:be:ef:00:XX". Within a short timespan
only 32 values are possible for \fIXX\fR.

When having overlapping networks and/or single IPs, this one with the
most specified netmask (count of 1's) takes precedence. IPs have a
netmask of 32. The behavior is unspecified when having overlapping
networks with the same count of 1's or when having duplicate IPs.

.\"
.SS PERFORMANCE
.\"
The lookup of single IPs has a complexity of O(log n) and this of
netmasks a complexity of O(n).

.\"
.SS SPECIAL MAC ADDRESSES
.\"
Beside the usual hex-octets-delimited-by-colons mac addresses,
ip-sentinel understands some special strings both on the commandline
and in the configuration file:
.TP
.BR LOCAL
expands to the mac-address of the used interface
.TP
.BR RANDOM
means a random mac-address which is newly calculated on every usage
.TP
.BR 802.1d
expands to 01:80:C2:00:00:00 which is the "Bridge Group Address".
.TP
.BR 802.3x
expands to 01:80:C2:00:00:01 which is the "IEEE Std. 802.3x Full
Duplex PAUSE operation". This MAC address will be blocked by a lot of
switches and will probably become the default in future versions.
.\"
.PP
The \fI802.*\fR addresses are having a special meaning for some
switches and packets having them as destination-address will be
dropped by the switch instead of flooding all ports. But it depends on
the used switch how/if these macs are honored.

Some switches can show an unexpected behavior are even crash if the
special \fI802.*\fR mac-addresses will be used.
.\"
.SS RANGES
.\"
Except in comments, it is possible to specify ranges everywhere in the
configuration file. These ranges are having the format "{from-to}" or
"{item1,item2,...,itemN}". The first format includes any number
beginning at "from" till "to" (inclusive), while the latter format
expands to the listed items only. The expansion happens on a
line-level and it is possible to use more than one range per line, so
that

.RS 0
192.168.0.{1-3}     0:0:0:0:0:1
.RS 0
192.168.1.{1,3}     0:0:0:0:0:2
.RS 0
192.168.{2,4}.{1-3} 0:0:0:0:0:3

is the same like writing

.RS 0
192.168.0.1         0:0:0:0:0:1
.RS 0
192.168.0.2         0:0:0:0:0:1
.RS 0
192.168.0.3         0:0:0:0:0:1
.RS 0
192.168.1.1         0:0:0:0:0:1
.RS 0
192.168.1.3         0:0:0:0:0:1
.RS 0
192.168.2.1         0:0:0:0:0:3
.RS 0
192.168.2.2         0:0:0:0:0:3
.RS 0
192.168.2.3         0:0:0:0:0:3
.RS 0
192.168.4.1         0:0:0:0:0:3
.RS 0
192.168.4.2         0:0:0:0:0:3
.RS 0
192.168.4.3         0:0:0:0:0:3

Because there can be created very much entries with a single line
(e.g. "{0-255}.{0-255}.{0-255}.{0-255}" would cover the entire IPv4
internet), ranges should be used sparely. When possible, large ranges
should be expressed with netmasks.

.\"
.SS EXAMPLE
.RS 0
0.0.0.0/0                 ## Block anything
.RS 0
!192.168.0.0/24           ## Allow IPs of the form 192.168.0.*
.RS 0
192.168.0.0               ## but block 192.168.0.0
.RS 0
192.168.0.1 a:b:c:d:e:f   ## use a special mac for 192.168.0.1
.RS 0
192.168.0.2 802.1d        ## and 01:80:C2:00:00:00 for 192.168.0.2

This setup will not send ARP-replies for the IPs 192.168.0.{3-255} but
when a host tries to use e.g. 169.254.145.213, ip-sentinel will tell
that this IP has a MAC of "de:ad:be:ef:00:XX".
.\"
.\" ====================
.\"
.SH SEE ALSO
.BR RFC\ 826 ,
.BR IEEE\ Std\ 802.1D
.\"
.\" ====================
.\"
.SH AUTHOR
Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>


ChangeLog
2004-06-17 15:03  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* NEWS, configure.ac, doc/index.html: version 0.10

2004-06-17 15:03  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/testsuite/simulate.c: added an explicit cast to avoid
	  compiler warnings

2004-06-16 12:43  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* ip-sentinel.spec.in: moved /etc/sysconfig/* files into -sysv
	  subpackage; they are not used for 'minit' anymore

2004-06-16 12:39  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* ip-sentinel.spec.in: updated minit filelist

2004-06-16 12:35  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/: Makefile.am, blacklist.c, util.c, util.h: added 'ensc_fmt'
	  support

2004-06-16 12:35  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/compat.h: added workaround for uint64* types in dietlibc
	  <0.25

2004-06-16 12:35  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/arpmessage.h: include <netinet/if_ether.h>

2004-06-16 12:34  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* contrib/ip-sentinel.params, src/fmt.h: initial checkin

2004-06-16 12:34  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* contrib/Makefile-files: removed minit-run scripts; it is done
	  more efficiently now...  added minit-params template

2004-06-16 12:34  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* contrib/ip-sentinel-run[DEAD]: removed minit-run scripts; it is
	  done more efficiently now...

2004-06-16 12:10  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/: Makefile-files, vector-zeroend.c, vector.h:
	  added Vector_zeroEnd() function

2004-06-16 12:06  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_fmt/: .cvsignore, Makefile-files, fmt-32.c, fmt-64.c,
	  fmt-internal.h, fmt-tai64n.c, fmt.h, fmt.hc, fmtx-32.c,
	  fmtx-64.c, fmtx.hc: moved fmt* functionality into the 'ensc_fmt'
	  module

2004-06-15 14:12  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* configure.ac: add '#include "compat.h"' to end of <config.h>

2004-06-15 14:11  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/: arguments.c, arguments.h, arpmessage.h, blacklist.c,
	  blacklist.h, ether.c, ip-sentinel.c, util.h, worker.c,
	  wrappers.h, testsuite/blacklist-check.c: do not include
	  "compat.h" anymore; this happens through <config.h> now

2004-06-15 14:05  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/testsuite/Makefile-files: cleanups for ensc_vector library

2004-06-15 14:04  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/: arguments.c, arguments.h, arpmessage.h, blacklist.c,
	  compat.h, util.c, worker.c: fixed build with recent dietlibc
	  (duplicate 'struct ethhdr' declarations)

2004-06-15 14:04  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/blacklist.h: fixed build with recent dietlibc (duplicate
	  'struct ethhdr' declarations) use 'ensc_vector' library instead
	  of 'vector*' files

2004-06-15 14:02  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/: vector.c[DEAD], vector.h[DEAD], vector.ic[DEAD]: use
	  'ensc_vector' library instead of 'vector*' files

2004-06-15 13:56  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/: Makefile.am, antidos.h: use 'ensc_vector' library instead
	  of 'vector*' files

2004-06-15 13:55  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* ip-sentinel.spec.in: removed 'Requires(...,...)' statements; they
	  are not working as expected :(

2004-03-24 02:12  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/.cvsignore: initial checkin

2004-03-24 02:11  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/testsuite/.cvsignore: added some standardfiles; I
	  do not want to wait until Savannah fixes its scripts and allows
	  access to CVSROOT/cvsignore again.

2004-03-10 03:30  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/: Makefile-files, testsuite/Makefile-files:
	  modernized the Makefiles; use '+=' where possible to remove
	  complexity from the toplevel Makefile.am

2004-02-17 20:51  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/testsuite/test1.c: make non-C99 compilers happy

2004-02-16 20:18  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/testsuite/test1.c: #undef NDEBUG

2004-02-10 13:24  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/testsuite/.cvsignore: initial checkin

2004-02-06 18:43  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/testsuite/: Makefile-files, test1.c: initial
	  checkin

2004-02-06 18:41  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/Makefile-files: added ensc_vector testsuite

2004-02-06 17:42  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/vector-insert.c: fixed it

2004-02-06 15:47  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* src/ensc_vector/: Makefile-files, vector-clear.c, vector-free.c,
	  vector-init.c, vector-insert.c, vector-internal.h,
	  vector-popback.c, vector-pushback.c, vector-resize.c,
	  vector-resizeinternal.hc, vector-search.c, vector-sort.c,
	  vector-unique.c, vector.h, vector.hc: initial checkin

2003-12-16 00:56  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* NEWS, configure.ac, doc/index.html, doc/ip-sentinel.man: version
	  0.9

2003-12-12 01:49  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* Makefile.am: - use ENSC_CHANGELOG - do not go into m4 anymore;
	  automake 1.8 provides this functionality   already

2003-12-12 01:49  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* configure.ac: - use ENSC_CHANGELOG - do not build m4/Makefile
	  anymore; automake 1.8 provides its   functionality already


src/ip-sentinel.c
// $Id: ip-sentinel.c,v 1.13 2003/05/26 21:49:22 ensc Exp $    --*- c++ -*--

// Copyright (C) 2002,2003 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//  

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "arguments.h"
#include "wrappers.h"
#include "util.h"
#include "parameters.h"
#include "blacklist.h"
#include "antidos.h"
#include "arpmessage.h"
#include "compat.h"
#include "ip-sentinel.h"

#include <assert.h>
#include <unistd.h>
#include <signal.h>
#include <net/if.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/wait.h>
#include <grp.h>

static struct ether_addr const	BCAST_MAC   = { { 255, 255, 255, 255, 255, 255 } };

volatile sig_atomic_t		child_count;
static volatile sig_atomic_t	do_reload;

static void sigChild(int);
static void sigHup(int);

void
sigChild(int sig UNUSED)
{
  while (waitpid(-1, 0, WNOHANG)>0) { if (child_count>0) --child_count; }
  signal(SIGCHLD, sigChild);
}

void
sigHup(int sig UNUSED)
{
  do_reload = 1;
  signal(SIGHUP, sigHup);
}

inline static void
adjustUserGroup(Arguments *arguments, uid_t *uid, gid_t *gid)
{
  struct passwd const *	pw_user;
  char *		err_ptr;

  assert(uid!=0 && gid!=0);
  
  *gid = static_cast(gid_t)(-1);
  *uid = strtol(arguments->user, &err_ptr, 0);
  if (arguments->user[0]!='\0' && *err_ptr=='\0') pw_user = getpwuid(*uid);
  else                                            pw_user = Egetpwnam(arguments->user);
  
  if (pw_user!=0) {
    *uid = pw_user->pw_uid;
    *gid = pw_user->pw_gid;
    if (arguments->chroot==0) arguments->chroot = pw_user->pw_dir;
  }

  if (arguments->group!=0) {
    *gid = strtol(arguments->group, &err_ptr, 0);
    if (arguments->group[0]=='\0' || *err_ptr!='\0')
      *gid = Egetgrnam(arguments->group)->gr_gid;
  }

  if (*gid==static_cast(gid_t)(-1)) {
    WRITE_MSGSTR(2, "Failed to determine gid; perhaps you should use the '-g' option. Aborting...\n");
    exit(1);
  }
}

static void
daemonize(Arguments *arguments)
{
  int			err_fd, out_fd, pid_fd;
  int			aux;
  uid_t			uid;
  gid_t			gid;
  pid_t			daemon_pid;

  err_fd = Eopen(arguments->errfile, O_WRONLY|O_CREAT|O_APPEND|O_NONBLOCK, 0700);
  out_fd = Eopen(arguments->logfile, O_WRONLY|O_CREAT|O_APPEND|O_NONBLOCK, 0700);
  pid_fd = (!arguments->do_fork ? -1 :
	    Eopen(arguments->pidfile, O_WRONLY|O_CREAT|O_TRUNC, 0755));

  adjustUserGroup(arguments, &uid, &gid);
  
  if (arguments->chroot != 0) {
    Echdir(arguments->chroot);
    Echroot(arguments->chroot);
  }

  Esetgroups(1, &gid);
  Esetgid(gid);
  Esetuid(uid);

  Eclose(0);

  aux = Eopen(arguments->ipfile, O_RDONLY, 0);
  Eclose(aux);

  if (arguments->do_fork) daemon_pid = Efork();
  else                    daemon_pid = 0;

  switch (daemon_pid) {
    case 0	:  break;
    default	:
      writeUInt(pid_fd, daemon_pid);
      Ewrite  (pid_fd, "\n", 1);
      assert(daemon_pid!=-1);
      exit(0);
      break;
  }

  if (arguments->do_fork) Esetsid();

  Edup2(out_fd, 1);
  Edup2(err_fd, 2);

  if (pid_fd!=-1) Eclose(pid_fd);
  Eclose(out_fd);
  Eclose(err_fd);
}

static int
getIfIndex(int fd, char const *iface_name)
{
  struct ifreq		iface;
  
  assert(iface_name!=0);
  memcpy(iface.ifr_name, iface_name, IFNAMSIZ);
  Eioctl(fd, SIOCGIFINDEX, &iface);
  if (iface.ifr_ifindex<0) {
    WRITE_MSGSTR(2, "No such interface '");
    WRITE_MSG   (2, iface_name);
    WRITE_MSGSTR(2, "'\n");
    exit(1);
  }

  return iface.ifr_ifindex;
}

static void
sendPacket(int s, int if_idx, struct ether_addr const *mac, in_addr_t const ip,
	   int count, long sleep_dur)
{
  ArpMessage d;
  struct sockaddr_ll	addr;

  memset(&d,    0, sizeof d);
  memset(&addr, 0, sizeof addr);

  d.padding[0] = 0x66;
  d.padding[1] = 0x60;
  d.padding[sizeof(d.padding)-2] = 0x0B;
  d.padding[sizeof(d.padding)-1] = 0x5E;
  
  d.data.ea_hdr.ar_hrd = htons(ARPHRD_ETHER);
  d.data.ea_hdr.ar_pro = htons(ETHERTYPE_IP);
  d.data.ea_hdr.ar_hln = ETHER_ADDR_LEN;
  d.data.ea_hdr.ar_pln = 4;
  d.data.ea_hdr.ar_op  = htons(ARPOP_REPLY);

  memcpy(d.data.arp_sha, mac->ether_addr_octet, sizeof(d.data.arp_sha));
  memcpy(d.data.arp_spa, &ip, sizeof(d.data.arp_spa));

  memcpy(d.data.arp_tha, BCAST_MAC.ether_addr_octet, sizeof(d.data.arp_tha));
  *(in_addr_t*)(&d.data.arp_tpa) = INADDR_ANY;

  memcpy(d.header.ether_dhost, BCAST_MAC.ether_addr_octet, sizeof(d.header.ether_dhost));
  memcpy(d.header.ether_shost, d.data.arp_sha,             sizeof(d.header.ether_shost));
  d.header.ether_type  = htons(ETH_P_ARP);

  addr.sll_family   = AF_PACKET;
  addr.sll_protocol = 0;
  addr.sll_ifindex  = if_idx;
  addr.sll_hatype   = 0;
  addr.sll_pkttype  = 0;
  addr.sll_halen    = ETHER_ADDR_LEN;
  memcpy(addr.sll_addr, d.data.arp_tha, ETHER_ADDR_LEN);

  while (count!=0) {
    int ret = TEMP_FAILURE_RETRY(sendto(s, &d, sizeof(d), 0,
					reinterpret_cast(struct sockaddr const *)(&addr), sizeof(addr)));

    FatalErrnoError(ret==-1, 1, "sendto()");
    
    if (count!=-1) --count;
    if (count!=0) sleep(sleep_dur);
  }
}

static void
handlePacket(int sock,  int if_idx, struct ether_addr const *mac,
	     struct ether_arp const *data)
{
  char		buffer[128];
  char *	buf_ptr = buffer;
  size_t	len     = sizeof(buffer);

  struct in_addr const		*spa = reinterpret_cast(struct in_addr const *)   (&data->arp_spa);
  struct in_addr const		*tpa = reinterpret_cast(struct in_addr const *)   (&data->arp_tpa);
  struct ether_addr const	*sha = reinterpret_cast(struct ether_addr const *)(&data->arp_sha);

  assert(spa!=0 && tpa!=0 && sha!=0);

    // Because this function will be called from a child-process and the logfile will not be locked,
    // try to make the write-operation as atomic as possible. Races can not prevented entirely since
    // the timestamp will be given out separately, but it is unlikely because concurrenting
    // child-processes will be started only after finishing some lengthy tasks (blacklist- and
    // antidos management)
  XSTRCAT(&buf_ptr, &len, ": Handle IP '");
  xstrcat(&buf_ptr, &len, inet_ntoa(*tpa));
  XSTRCAT(&buf_ptr, &len, "' requested by '");
  xstrcat(&buf_ptr, &len, inet_ntoa(*spa));
  XSTRCAT(&buf_ptr, &len, "' [");
  xstrcat(&buf_ptr, &len, ether_ntoa(sha));
  XSTRCAT(&buf_ptr, &len, "]\n");

  writeMsgTimestamp(1);
  write(1, buffer, sizeof(buffer)-len);

  sendPacket(sock, if_idx, mac, tpa->s_addr, 1, 0);
  usleep(1500000);
  sendPacket(sock, if_idx, mac, tpa->s_addr, 3, 2);
}

static void
handleMessage(int sock, int if_idx, struct ether_addr const *mac, struct ether_arp const *data)
{
  static int		error_count =0;
  pid_t			pid;
  sigset_t		block_set, old_set;

  while (child_count>=MAX_CHILDS) {
    writeMsgTimestamp(2);
    WRITE_MSGSTR(2, ": Too much children (");
    writeUInt(2, child_count);
    WRITE_MSGSTR(2, ") forked; sleeping a while...\n");
    sleep(1);
  }

  sigfillset(&block_set);
  sigprocmask(SIG_BLOCK, &block_set, &old_set);

  pid = fork();
  switch (pid) {
    case -1	:
      perror("fork()");
      writeMsgTimestamp(2);
      WRITE_MSGSTR(2, ": fork() failed");
      ++error_count;
      if (error_count>MAX_ERRORS) {
	WRITE_MSGSTR(2, "aborting...");
	exit(1);
      }
      break;
      
    case 0	:
      handlePacket(sock, if_idx, mac, data);
      exit(0);

    default	:
      ++child_count;
      error_count = 0;
      break;
  }

  sigprocmask(SIG_SETMASK, &old_set, 0);
}

static void NORETURN
run(int sock, int if_idx, char const *filename) 
{
  BlackList			cfg;
  AntiDOS			anti_dos;
  int				error_count = 0;
  struct sockaddr_ll		addr;
  socklen_t			from_len = sizeof(addr);
  char				buffer[4096];
  ArpMessage const * const	msg    = reinterpret_cast(ArpMessage const *)(buffer);
  struct in_addr const	*	src_ip = reinterpret_cast(struct in_addr const *)(msg->data.arp_spa);
  unsigned int			oversize_sleep = 1;

  memset(&addr, 0, sizeof(addr));
  addr.sll_family  = AF_PACKET;
  addr.sll_ifindex = if_idx;
  
  BlackList_init(&cfg, filename);
  AntiDOS_init(&anti_dos);

  while (true) {
    size_t			size;
    int				arp_count;
    struct ether_addr const	*mac;
    struct ether_addr		mac_buffer;
    
    AntiDOS_update(&anti_dos);
    size = TEMP_FAILURE_RETRY(recvfrom(sock, buffer, sizeof buffer, 0,
				       (struct sockaddr *)(&addr), &from_len));

    if (static_cast(ssize_t)(size)==-1) {
      ++error_count;
      if (error_count>MAX_ERRORS) {
	perror("recvfrom()");
	exit(1);
      }

      continue;
    }

    error_count = 0;

    if (ntohs(addr.sll_protocol)!=ETHERTYPE_ARP)      continue;
    if (ntohs(msg->data.ea_hdr.ar_op)!=ARPOP_REQUEST) continue;

    if (!do_reload) BlackList_softUpdate(&cfg);
    else {
      BlackList_update(&cfg);
      do_reload = false;
    }

    mac = BlackList_getMac(&cfg, *reinterpret_cast(struct in_addr *)(msg->data.arp_tpa), &mac_buffer);
    if (mac==0) continue;

    assert(src_ip!=0);
    if (!AntiDOS_isOversized(&anti_dos)) oversize_sleep = 1;
    else {
      writeMsgTimestamp(2);
      WRITE_MSGSTR(2, ": Too much requests from too much IPs; last IP was ");
      writeIP     (2, *src_ip);
      WRITE_MSGSTR(2, "\n");

      sleep(oversize_sleep);
      oversize_sleep = MIN(oversize_sleep+1, 10);
      continue;
    }

    arp_count = AntiDOS_registerIP(&anti_dos, *src_ip);

    if (isDOS(arp_count)) {
      writeMsgTimestamp(2);
      WRITE_MSGSTR(2, ": Too much requests from ");
      writeIP     (2, *src_ip);
      WRITE_MSGSTR(2, "; DOS-measurement was ");
      writeUInt   (2, arp_count);
      WRITE_MSGSTR(2, "\n");

      continue;
    }


    handleMessage(sock, if_idx, mac, &msg->data);
  }
}

int
main(int argc, char *argv[])
{
  Arguments		arguments;
  int			sock;
  int			if_idx;
  
  parseOptions(argc, argv, &arguments);

  sock   = Esocket(PF_PACKET, SOCK_RAW, htons(ETH_P_ARP));
  if_idx = getIfIndex(sock, arguments.iface);
  daemonize(&arguments);

  signal(SIGCHLD, sigChild);
  signal(SIGHUP,  sigHup);

  run(sock, if_idx, arguments.ipfile);
}
